packages/compass-e2e-tests/smoke-test.ts

```diff
@@ -0,0 +1,338 @@
+  'linux_deb_filename',
+  'linux_tar_filename',
+] as const;
+type UbuntuBuildInfo = Record<typeof ubuntuFilenameKeys[number], string>;
+
+const rhelFilenameKeys = ['linux_rpm_filename', 'rhel_tar_filename'] as const;
+type RHELBuildInfo = Record<typeof rhelFilenameKeys[number], string>;
+
+function buildInfoIsWindows(buildInfo: any): buildInfo is WindowsBuildInfo {
+  for (const key of windowsFilenameKeys) {
+    if (!buildInfo[key]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function buildInfoIsOSX(buildInfo: any): buildInfo is OSXBuildInfo {
+  for (const key of osxFilenameKeys) {
+    if (!buildInfo[key]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function buildInfoIsUbuntu(buildInfo: any): buildInfo is UbuntuBuildInfo {
+  for (const key of ubuntuFilenameKeys) {
+    if (!buildInfo[key]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function buildInfoIsRHEL(buildInfo: any): buildInfo is RHELBuildInfo {
+  for (const key of rhelFilenameKeys) {
+    if (!buildInfo[key]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+type Package = {
+  name: string;
+  filepath: string;
+};
+
+function getFilteredPackages(
+  compassDir: string,
+  buildInfo: any,
+  config: PackageFilterConfig
+): Package[] {
+  let names: string[] = [];
+
+  if (config.isWindows) {
+    if (!buildInfoIsWindows(buildInfo)) {
+      throw new Error('missing windows package keys');
+    }
+    names = windowsFilenameKeys.map((key) => buildInfo[key]);
+  } else if (config.isOSX) {
+    if (!buildInfoIsOSX(buildInfo)) {
+      throw new Error('missing osx package keys');
+    }
+    names = osxFilenameKeys.map((key) => buildInfo[key]);
+  } else if (config.isRHEL) {
+    if (!buildInfoIsRHEL(buildInfo)) {
+      throw new Error('missing rhel package keys');
+    }
+    names = rhelFilenameKeys.map((key) => buildInfo[key]);
+  } else if (config.isUbuntu) {
+    if (!buildInfoIsUbuntu(buildInfo)) {
+      throw new Error('missing ubuntu package keys');
+    }
+    names = ubuntuFilenameKeys.map((key) => buildInfo[key]);
+  }
+
+  const extension = config.extension;
+
+  return names
+    .filter((name) => !extension || name.endsWith(extension))
+    .map((name) => {
+      return {
+        name,
+        filepath: path.join(compassDir, 'dist', name),
+      };
+    });
+}
+
+async function downloadFile(url: string, targetFile: string): Promise<void> {
+  return await new Promise((resolve, reject) => {
+    https
+      .get(url, (response) => {
+        const code = response.statusCode ?? 0;
+
+        if (code >= 400) {
+          return reject(new Error(response.statusMessage));
+        }
+
+        // handle redirects
+        if (code > 300 && code < 400 && !!response.headers.location) {
+          return resolve(downloadFile(response.headers.location, targetFile));
+        }
+
+        // save the file to disk
+        const fileWriter = createWriteStream(targetFile).on('finish', () => {
+          resolve();
+        });
```

Comment:
Perhaps you'd want to register an 'error' listener here as well (calling `reject`), in case the write stream fails?

Reply:
I suppose that won't hurt. I find streams such a footgun. There really should be builtin utilities somewhere for these cases. We use the `download` npm package in a few other places but it hasn't been published to in 5 years, supports no ESM, no TypeScript.. with no obvious perfect alternative I could quickly find.

Reply:
How about using the `finished` function on the builtin `node:stream/promises` module?

https://nodejs.org/api/stream.html#streamfinishedstream-options

Reply:
Or perhaps [this answer](https://stackoverflow.com/a/77210219) on stack-overflow:

```ts
import { writeFile } from 'node:fs/promises';
import { Readable } from 'node:stream';

const response = await fetch(url);
const stream = Readable.fromWeb(response.body);
await writeFile(targetFile, stream);
```

Reply:
Late to the party, but the best-practice generic way here (i.e. not relying on this particularity of the `writeFile()` API) would be to use `stream.promises.pipeline(response, createWriteStream(targetFile));` – that already takes care of error handling and waiting for the operation to finish for you.